explainaboard/metrics/metric.py

```diff
@@ -269,44 +269,107 @@ def calc_stats_from_data(
         evaluation metric can be calculated later. In the simplest form, this is just
         the evaluation metric value for each example.
 
-        :param true_data: gold-standard data
-        :param pred_data: predicted data
-        :param config: a configuration to over-ride the default for this object
-        :return: a numpy array of shape [len(true_data), X] where X=1 in the simplest
-            case of decomposable eval metrics
+        Args:
+            true_data: gold-standard data
+            pred_data: predicted data
+            config: a configuration to over-ride the default for this object
+
+        Returns:
+            A numpy array of shape [len(true_data), X] where X=1 in the simplest case of
+            decomposable eval metrics
         """
         ...
 
-    def aggregate_stats(self, stats: MetricStats) -> np.ndarray:
+    @final
+    def aggregate_stats(
+        self, stats: MetricStats
+    ) -> np.ndarray[tuple[int], Any] | np.ndarray[tuple[int, int], Any]:
         """Aggregate sufficient statistics from multiple examples into a single example.
 
         Args:
             stats: stats for every example
 
         Returns:
-            Aggregated stats
+            Aggregated stats. Shape must be:
+                - Non-batched data: [num_stats]
+                - Batched data: [batch_size, num_stats]
         """
+        result = self._aggregate_stats(stats)
+
+        if stats.is_batched():
+            stats_shape = stats.get_batch_data().shape
+            result_shape: tuple[int, ...] = (stats_shape[0], stats_shape[2])
+        else:
+            result_shape = (stats.num_statistics(),)
```

Comment:
@odashi Could you tell me why this is the case? i.e., Must one of the dimensions of aggregated result be the same as `num_statistics`?

Reply:
Yes (see docstring). This function expects to process only aggregating the same statistics ([[A,B,C], [A',B',C']] -> [A'', B'', C''])

Reply:
I see. 

So there are two types of `aggregation`s,
* (1) aggregation over different samples
* (2) aggregation over different statistics 

Based on my understanding, the current aggregation function says: (1) must happen before (2).
I'm not sure if this is too strong. 
For example, there are some scenarios, where (2) should happen first (different statistics should be composed together to calculate a metric value)

What do you think of this situation?

Reply:
Hi @pfliu-nlp : just to clarify, are you referring to the following?
* (1) `aggregate_stats()`
* (2) `calc_metric_from_aggregate()`

If that's the case, `calc_metric_from_aggregate()` can also be applied to batched statistics, so it could theoretically be called before `(1)` as well to calculate metrics on multiple examples as well.

Reply:
Some confusion comes from the fact that the interface takes and returns ndarrays directly.
More concretely, we need to make sure that:

1. `aggregate_stats` takes the concept *Stats* and returns the concept *Aggregate*
2. `calc_metric_from_aggregate` takes the concept *Aggregate* and retruns the concept *Result*

*Stats*, *Aggregate* and *Result* are represented as `ndarray`s (though *Stats* are obtained from `MetricStats`), but they don't share any underlying semantics.
According to this understanding, there is no chance that 2. precedes 1 as @neubig said too.

Reply:
More and more concretely:

- $\mathrm{aggregate\\_stats}: List[Stats] \rightarrow Stats \ \ OR \ \ Batch[List[Stats]] \rightarrow Batch[Stats]$
- $\mathrm{calc\\_metric\\_from\\_aggregate}: Stats \rightarrow Result \ \ OR \ \ Batch[Stats] \rightarrow Batch[Result]$

Reply:
Yes, I like this formulation.